File: 2020/Swift/06-CustomCustoms/CustomCustoms.swift
#!/usr/bin/swift

// Custom Customs
// https://adventofcode.com/2020/day/6

import Foundation

func readFile(named name: String) -> [String] {
    let currentDirectoryURL = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
    let fileURL = URL(fileURLWithPath: name + ".txt", relativeTo: currentDirectoryURL)
    guard let content = try? String(contentsOf: fileURL, encoding: String.Encoding.utf8) else {
        print("Unable to read input file \(name)")
        print("Current directory: \(currentDirectoryURL)")
        return []
    }
    return content.components(separatedBy: .newlines)
}

func questionsAnyoneAnswered(declarations: ArraySlice<String>) -> Set<Character> {
    declarations.reduce([]) { (result, declaration) -> Set<Character> in
        result.union(declaration)
    }
}

func questionsAllAnswered(declarations: ArraySlice<String>) -> Set<Character> {
    declarations.reduce(Set<Character>("abcdefghijklmnopqrstuvwxyz")) { (result, declaration) -> Set<Character> in
        result.intersection(declaration)
    }
}

let declarationsStrings = readFile(named: "06-input")
let groups = declarationsStrings.split(separator: "")

let anyAnswers = groups.map { questionsAnyoneAnswered(declarations: $0) }
let sumOfAnyQuestionCounts = anyAnswers.map { $0.count }.reduce(0, +)
print("The sum of the counts of any yes answers is \(sumOfAnyQuestionCounts)")

let allAnswers = groups.map { questionsAllAnswered(declarations: $0) }
let sumOfAllQuestionCounts = allAnswers.map { $0.count }.reduce(0, +)
print("The sum of the counts of all yes answers is \(sumOfAllQuestionCounts)")

File: 2020/Python/06-CustomCustoms/CustomCustoms.py
#!/usr/bin/env python3

# Custom Customs
# https://adventofcode.com/2020/day/6

from functools import reduce

def read_file(name):
    file = open(name)
    return list(file.readlines())

def parse_groups(lines):
    declarations = []
    declaration = []
    for line in lines:
        line = line.strip()
        if not line:
            declarations.append(declaration)
            declaration = []
            continue
        declaration.append(line)
    if declaration:
        declarations.append(declaration)
    return declarations

def questionsAnyoneAnswered(declarations):
    return reduce(lambda result, value: result.union(value), declarations, set())

def questionsAllAnswered(declarations):
    return reduce(lambda result, value: result.intersection(value), declarations, set("abcdefghijklmnopqrstuvwxyz"))

groups = parse_groups(read_file('06-input.txt'))

any_answers = map(lambda x: questionsAnyoneAnswered(x), groups)
sum_of_any_question_counts = sum(map(lambda x: len(x), any_answers))
print("The sum of the counts of any yes answers is {0}".format(sum_of_any_question_counts))

all_answers = map(lambda x: questionsAllAnswered(x), groups)
sum_of_all_question_counts = sum(map(lambda x: len(x), all_answers))
print("The sum of the counnts of all yes answers is {0}".format(sum_of_all_question_counts))
